coerce and rotate pvgis TMY data to desired tz and year

mikofski opened the pull request:
- add private function `_coerce_and_roll_pvgis_tmy()`
- add `roll_utc_offset` and `coerce_year` params to docstring for `get_pvgis_tmy`
- call private function if `roll_utc_offset` is not `None` or `coerce_year` is not `None`

discussed in https://github.com/pvlib/pvlib-python/discussions/1528#discussioncomment-9177537

<!-- Thank you for your contribution! The following items must be addressed before the code can be merged. Please don't hesitate to ask for help if you're unsure of how to accomplish any of the items. Feel free to remove checklist items that are not relevant to your change. -->

 - [x] Closes #2139 
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [ ] ~Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.~
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

<!-- Brief description of the problem and proposed solution (if not already fully described in the issue linked to above): -->

mikofski commented:
Tested in this Colab notebook: [`test_coerce_pvgis.ipynb`](https://colab.research.google.com/drive/1pibVTd02QUEJ45oShWuaqzywqviaxWFo?usp=sharing).

cwhanse commented:
Showing my ignorance of PVGIS: does it always return 8760 timestamps? 

If PVGIS does return 8760, what should I expect if a user coerces to a leap year? I assume they would get a DatetimeIndex with Feb 29 missing.

Second question: if I rotate to UTC-10 (`utc_offset=-10`), what should I expect as the result? An index of length 8750, or an index of length 8760 but with the last 10 hours in the following year? If its 8760, what data are in the last 10 hours?

mikofski commented:
- `get_pvgis_tmy()` always returns 8760 in utc
- yes, you are correct. If coerced to a leap year then Feb 29 is skipped, this matches other functions in `iotools` with this argument
- if you rotate to utc offset of -10 then the output will be a 8760 starting at midnight of the forced year, the first 10 rows are “rotated” to the bottom, because in utc-10 they are from 2am to 11pm on December 31st. The tests check the output to see that the first timestamp is January 1st at midnight, and the last timestamp is December 31st at 11pm

Where is utc-10 is that Hawaii?

cwhanse commented:
> Where is utc-10 is that Hawaii?

Dunno I just picked a number for illustration.

cwhanse commented:
> if you rotate to utc offset of -10 then the output will be a 8760 starting at midnight of the forced year, the first 10 rows are “rotated” to the bottom, because in utc-10 they are from 2am to 11pm on December 31st. 

Ah. Basically we assume a TMY year is periodic, thus every hour's data is defined in the rotated index.

adriesse commented on pvlib/iotools/pvgis.py:
```diff
@@ -390,9 +391,32 @@ def read_pvgis_hourly(filename, pvgis_format=None, map_variables=True):
     raise ValueError(err_msg)
 
 
+def _coerce_and_rotate_pvgis(pvgis_data, tz, year):
```
```suggestion
def _coerce_and_rotate_tmy(tmy_data, tz):
```
Should only be used for tmy, right?  Source probably doesn't matter.

Does pvgis put out Feb 29?  I hope not.  I think it would be ok to just use 1990.

mikofski replied:
Yes pvgis leaves out Feb 29 always even if month is 2012, a leap year

mikofski replied:
changed name of private function to `_coerce_and_rotate_pvgis_tmy`

adriesse commented on pvlib/iotools/pvgis.py:
```diff
@@ -424,6 +448,12 @@ def get_pvgis_tmy(latitude, longitude, outputformat='json', usehorizon=True,
         time in seconds to wait for server response before timeout
+    utc_offset: int, default 0
+        Use to specify a timezone other than the default UTC zero. Will force
+        year to ``coerce_year`` if not zero.
```
```suggestion
        year to 1990 if not zero.
```

mikofski replied:
I prefer to keep function to coerce years, matches other functions in iotools

adriesse replied:
No problem, I hadn't looked at the other functions.

adriesse commented on pvlib/iotools/pvgis.py:
```diff
@@ -424,6 +448,12 @@ def get_pvgis_tmy(latitude, longitude, outputformat='json', usehorizon=True,
+    utc_offset: int, default 0
+        Use to specify a timezone other than the default UTC zero. Will force
+        year to ``coerce_year`` if not zero.
+    coerce_year: int, default 1990
```
```suggestion
    coerce_year: boolean
```

mikofski replied:
doesn't match coerce year in other iotools which lets user force any year

adriesse commented on pvlib/iotools/pvgis.py:
```diff
@@ -424,6 +448,12 @@ def get_pvgis_tmy(latitude, longitude, outputformat='json', usehorizon=True,
+        Use to specify a timezone other than the default UTC zero. Will force
+        year to ``coerce_year`` if not zero.
+    coerce_year: int, default 1990
+        Use to force indices to desired year. Ignored if ``utc_offset`` is
+        zero.
```
```suggestion
        Used to force indices to 1990. Set to True if ``utc_offset`` is
        non-zero.
```

mikofski replied:
this doesn't follow the usage in other functions, and I think that pvgis main issue is the `utc_offset` so my preference is to keep both args and allow user to coerce years like other `iotools` do

adriesse commented on pvlib/iotools/pvgis.py:
```diff
@@ -390,9 +391,32 @@ def read_pvgis_hourly(filename, pvgis_format=None, map_variables=True):
+    """
+    # assert tz//1 == tz
+    tzname = f'Etc/GMT{-tz:+d}'
+    # check if February is leap year
+    feb1 = pvgis_data[pvgis_data.index.month == 2].index.year[0]
+    if calendar.isleap(feb1):
+        # replace Feb year with a non-leap year
+        pvgis_data.index = [
+            ts.replace(year=1990) if ts.month == 2 else ts
+            for ts in pvgis_data.index]
```
Can skip all this if we reduce the options.

adriesse replied:
If you always coerce the year to 1990 there will be no exception.

mikofski replied:
i think the order of operations has to be
1. Sanitize February
2. Rotate hours
3. Concert time zone
4. replace years

If I try to coerce years first then when I convert time zone I get two years.

i can try to experiment with the order but this was what seemed to work best. Let me play around with it some more and maybe I can get rid of step 1. Although I do really like the option to coerce to any year not just 1990.

mikofski replied:
I found a way to make it work without sanitizing February and with `np.roll` by replacing the year and time zone at the same time. Any other way always returns two years.

adriesse commented on pvlib/iotools/pvgis.py:
```diff
@@ -390,9 +391,32 @@ def read_pvgis_hourly(filename, pvgis_format=None, map_variables=True):
+        pvgis_data.index = [
+            ts.replace(year=1990) if ts.month == 2 else ts
+            for ts in pvgis_data.index]
+    pvgis_data = pd.concat(
+        [pvgis_data.iloc[-tz:], pvgis_data.iloc[:-tz]],
+        axis=0).tz_convert(tzname)
```
I kind of liked the `np.roll()` solution but in python there are always _so_ many ways to achieve the same result!

mikofski replied:
I don’t recall why I chose this approach but I can test if `np.roll` works better

mikofski replied:
it works!

cwhanse commented on pvlib/tests/iotools/test_pvgis.py:
```diff
@@ -435,6 +435,22 @@ def _compare_pvgis_tmy_basic(expected, meta_expected, pvgis_data):
+    jan1_midnight2021 = pd.Timestamp('2021-01-01 00:00:00', tz='Etc/GMT-2')
+    dec31_midnight2021 = pd.Timestamp('2021-12-31 23:00:00', tz='Etc/GMT-2')
+    assert pvgis_data[0].index[0] == jan1_midnight2021
+    assert pvgis_data[0].index[-1] == dec31_midnight2021
```
Can we also check that a few values are aligned with the index correctly?

mikofski replied:
I checked noon for every month with default and coerced year. Note I conflated a few other updates to the test:
- check that UTC time zone is unchanged if using defaults
- update Turin is actually CET which is UTC+1, not +2, oops! I got confused because currently it is CEST, or summertime which _is_ +2, 😆 ha-ha
- use some variables to be DRY to replace some WET & hardcoded args like the coerced year used 3x, time zone used 4x, and the "Etc/GMT" tz name also used 4x
- I didn't test the combo of `utc_offset` is zero, and `coerce_year` is set - should do nothing - do we need this test?

kandersolar commented on pvlib/iotools/pvgis.py:
```diff
@@ -424,6 +441,12 @@ def get_pvgis_tmy(latitude, longitude, outputformat='json', usehorizon=True,
         time in seconds to wait for server response before timeout
+    utc_offset: int, default 0
+        Use to specify a timezone other than the default UTC zero. Will force
+        year to ``coerce_year`` if not zero.
```
This parameter names and description don't make any mention of rolling the data.  I think at least the parameter description should explain the rolling.  Ideally the parameter name would somehow reflect it as well, although I'm having trouble coming up with a good name.  Somehow `roll_according_to_utc_offset` doesn't quite roll off the tongue...

adriesse replied:
I was going to leave good enough alone, but since we have more suggestions: I would roll automatically based on longitude.

mikofski replied:
I can add explanation of rolling and maybe somewhere we should add description that TMY is always expected to start and end at midnight on Jan 1/December 31?

mikofski replied:
>roll automatically based on longitude.

I don’t want to make rolling mandatory. Using longitude is clever, but not all time zones are strictly longitudinal, and I’ve had issues using that approach.

adriesse replied:
No problem.

adriesse replied:
> Somehow `roll_according_to_utc_offset` doesn't quite roll off the tongue...

How about `align_to_local_noon`?

mikofski replied:
@adriesse  I changed it to `roll_utc_offset` is that okay? It's both a verb `roll` and a number `utc_offset`. I like your suggestion, and thought my tests should probably have check transit times, which is the real test of whether it's in the right time zone, but that was a bridge too far.

adriesse replied:
It's all good.  No strong feelings about any of this.

kandersolar reviewed (commented):
FYI, the PSM3 TMY API does this "roll over" also: https://assessingsolar.github.io/unofficial-psm3-userguide/pages/rollover.html

adriesse commented:
> FYI, the PSM3 TMY API does this "roll over" also: https://assessingsolar.github.io/unofficial-psm3-userguide/pages/rollover.html

TMY methods have ways of merging between months from different years.  Ideally this would be done here.

mikofski commented on pvlib/iotools/pvgis.py:
```diff
@@ -510,6 +533,9 @@ def get_pvgis_tmy(latitude, longitude, outputformat='json', usehorizon=True,
         data = data.rename(columns=VARIABLE_MAP)
 
+    if utc_offset != 0:
+        data = _coerce_and_roll_pvgis_tmy(data, utc_offset, coerce_year)
```
I did recognize that forcing year could be useful by itself (eg when utc is zero)  but figured users could more easily replace the index with time series of their choice since no rolling is required. Eg:
```python
data.index = pd.date_range(start='year-01-01 01:00', freq='T', …)
```
I didn’t want to make `utc_offset` be either boolean or integer, because I’m afraid it’s a risky design that later will cause unexpected problems, eg zero is also false. I also didn’t want to add a 3rd new argument, so that’s why I left it like this.

kandersolar replied:
Would something like the following work?

```python
def get_pvgis_tmy(..., roll_offset=0, coerce_year=1990):
    """
    Parameters
    ----------
    ...
    roll_offset : int, default 0
        A UTC offset in hours.  If provided, the dataframe will be "rolled"
        so that data begins at midnight, blah blah
        It is likely desirable to use this parameter in combination with
        ``coerce_year``.
    coerce_year : int, default 1990
        blah blah
    """

    if roll_offset != 0:
        tz = pytz.timezone(f'Etc/GMT{-roll_offset:+d}')
        data = data.tz_convert(tz)
        data = pd.DataFrame(np.roll(data, roll_offset, axis=0),
                            columns=data.columns,
                            index=np.roll(data.index, roll_offset))

    if coerce_year:
        data.index = [x.replace(year=1990) for x in data.index]
```

mikofski replied:
I don’t think this will work. When I tried this variant I got an error that February 29 is not a date. The `tz_convert()` changed the hour which made March 1st at midnight become 1am and made midnight now February 29th. I don’t want to have to sanitize the February year PVGIS chooses before converting. Look back at 31800e9 and d0bd0a0.

kandersolar replied:
Argh, good point.  Shame on me for not actually trying it.  

I still think it's strange for `coerce_year` to only work when `utc_offset` is nonzero.  Does your code still work if it runs `_coerce_and_roll_pvgis_tmy` even when `utc_offset` is zero?  Seems like it should...

mikofski replied:
I can make it work - give me a second

adriesse replied:
> Argh, good point. Shame on me for not actually trying it.
> 
> I still think it's strange for `coerce_year` to only work when `utc_offset` is nonzero. Does your code still work if it runs `_coerce_and_roll_pvgis_tmy` even when `utc_offset` is zero? Seems like it should...

`align_to_local_noon` might be less cryptic.

mikofski replied:
@kandersolar ready for review -- see my comment (updated) and my last 2 commits.

kandersolar commented on pvlib/iotools/pvgis.py:
```diff
@@ -424,6 +441,13 @@ def get_pvgis_tmy(latitude, longitude, outputformat='json', usehorizon=True,
+        Use to specify a time zone other than the default UTC zero and roll
+        dataframe by ``utc_offset`` so that it starts at midnight on January
+        1st. If not zero, will also force year to ``coerce_year``.
+    coerce_year: int, default 1990
```
Let's do `coerce_year=None` instead, for consistency with the other pvlib functions that have `coerce_year` (`read_epw`, `read_tmy3`).  

If we do have a default year, that would be a change in behavior that could warrant a deprecation period...

mikofski replied:
I think if we make both defaults `None`I can check identity, add a comment to avoid implicit booleaness, and then make it possible to coerce the year even for utc zero. I will send an update later today.

I prefer `utc_offset` over `roll_offset` because this function converts the time zone to a new utc offset, which is not explained in “roll offset” while utc offset is a well known concept. How about `roll_utc_offset` which unfortunately sounds like a verb, but makes some sense if I let the default be `None`?

kandersolar commented on pvlib/iotools/pvgis.py:
```diff
@@ -424,6 +443,13 @@ def get_pvgis_tmy(latitude, longitude, outputformat='json', usehorizon=True,
         base url of PVGIS API, append ``tmy`` to get TMY endpoint
     timeout : int, default 30
         time in seconds to wait for server response before timeout
+    roll_utc_offset: int, default None
+        Use to specify a time zone other than the default UTC zero and roll
+        dataframe by ``roll_utc_offset`` so it starts at midnight on January
+        1st. Ignored if ``None``, otherwise will force year to ``coerce_year``.
+    coerce_year: int, default None
+        Use to force indices to desired year. Will default to 1990 if
+        ``coerce_year`` is ``None``, but ``roll_utc_offset`` is not ``None``.
```
```suggestion
    roll_utc_offset: int, optional
        Use to specify a time zone other than the default UTC zero and roll
        dataframe by ``roll_utc_offset`` so it starts at midnight on January
        1st. Ignored if ``None``, otherwise will force year to ``coerce_year``.
    coerce_year: int, optional
        Use to force indices to desired year. Will default to 1990 if
        ``coerce_year`` is not specified, but ``roll_utc_offset`` is specified.
```

For #1574

kandersolar commented on docs/sphinx/source/whatsnew/v0.11.1.rst:
```diff
@@ -16,6 +16,10 @@ Enhancements
 * Add new parameters for min/max absolute air mass to
   :py:func:`pvlib.spectrum.spectral_factor_firstsolar`.
   (:issue:`2086`, :pull:`2100`)
+* Add ``utc_offset`` and ``coerce_year`` arguments to
```
```suggestion
* Add ``roll_utc_offset`` and ``coerce_year`` arguments to
```

kandersolar reviewed (approved):
LGTM.  We may want to add similar functionality for other TMY datasets in the future, so I think it's worth @AdamRJensen taking a look too if he has the time.

AdamRJensen commented:
> > LGTM. We may want to add similar functionality for other TMY datasets in the future,
> 
> LGTM2
> 
> My perception is that this is fixing a flaw in the PVGIS service. Hopefully other TMY sources align on local time.

The PVGIS team is currently reworking the entire process, so this is a good time to make requests about features and provide feedback. Tagging @NikosAlexandris

AdamRJensen commented on pvlib/iotools/pvgis.py:
```diff
@@ -510,6 +536,16 @@ def get_pvgis_tmy(latitude, longitude, outputformat='json', usehorizon=True,
+    if roll_utc_offset is not None or coerce_year is not None:
+        # XXX: be explicit, test for identity not implicit booleaness
+        # roll_utc_offset None but coerce_year isn't, set year with utc zero
+        if roll_utc_offset is None:  # XXX: None and zero are both False
+            roll_utc_offset = 0
```
Is this really a desired behavior? Alternatively would be to raise a ValueError if roll_utc_offset is missing

mikofski replied:
I think it's important because it allows user to coerce year, but leave TZ alone - especially useful in GMT. @kandersolar requested this in https://github.com/pvlib/pvlib-python/pull/2138#discussion_r1696869191. I reworked `_coerce_and_roll_tmy()` to accept either `None` or zero as UTC, so I removed this from `get_pvgis_tmy()` for simplicity.

mikofski replied:
Done in fb2813d.

AdamRJensen replied:
I guess a user could pass in a non-utc localized index (assuming the timezone would get picked up by the function) but then it would get overridden if the timezone wasn't separately given as an input?

mikofski replied:
`_coerce_and_roll_tmy()` is a private function in the `pvgis.py` module that only gets called by `get_pvgis_tmy()`, and PVGIS TMY are always UTC. This is the purpose of #2139 which is to localize TMY dataset from PVGIS and roll it so it starts at midnight.

IMHO private functions are undocumented, so responsibility is the at the user’s own risk if `_coerce_and_roll_tmy()` is passed a localized index.

AdamRJensen replied:
Right right, I'm not requesting changes just wanted to clarify my understanding.

AdamRJensen commented on pvlib/iotools/pvgis.py:
```diff
@@ -510,6 +536,16 @@ def get_pvgis_tmy(latitude, longitude, outputformat='json', usehorizon=True,
         data = data.rename(columns=VARIABLE_MAP)
 
+    if roll_utc_offset is not None or coerce_year is not None:
+        # XXX: be explicit, test for identity not implicit booleaness
```
what does these "xxx" mean?

mikofski replied:
I use it as a warning to beware or pay attention or to flag problematic code. I used it here because I was worried about implicit booleaness, especially that `None` and 0 are both false, which could be a problem because zero is a valid UTC offset, but evaluates the same as `None` even though we want the opposite outcome.

Various places define it differently, but I disagree that's it's synonymous with or related to `FIXME:` or `TODO:`
- [wikipedia > computer programming comments > tags](https://en.wikipedia.org/wiki/Comment_(computer_programming)#Tags)
- [SO what does `XXX` mean?](https://softwareengineering.stackexchange.com/questions/65467/what-does-xxx-mean-in-a-comment)
- [SO what is the meaning of `XXX`?](https://stackoverflow.com/questions/1452934/what-is-the-meaning-of-xxx-in-code-comments)
- [SO best practices rule 9](https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/)

mikofski replied:
I removed the warning comments now that I handle `tz=None` in `_coerce_and_roll_tmy()` instead of here.

AdamRJensen reviewed (approved):
I tested it and it seems to work great 🥳 

One thing I noticed was that the index name is maintained as "time(UTC)" when utc_roll_offset is None and is removed when utc_roll_offset is specified. Should this be tested? Maybe it doesn't matter

mikofski commented on pvlib/iotools/pvgis.py:
```diff
@@ -390,9 +390,28 @@ def read_pvgis_hourly(filename, pvgis_format=None, map_variables=True):
     raise ValueError(err_msg)
 
 
+def _coerce_and_roll_pvgis_tmy(pvgis_data, tz, year):
```
Agreed, removed "pvgis" from name in 91c169e.

mikofski commented on pvlib/iotools/pvgis.py:
```diff
@@ -424,6 +443,13 @@ def get_pvgis_tmy(latitude, longitude, outputformat='json', usehorizon=True,
+        dataframe by ``roll_utc_offset`` so it starts at midnight on January
+        1st. Ignored if ``None``, otherwise will force year to ``coerce_year``.
+    coerce_year: int, optional
+        Use to force indices to desired year. Will default to 1990 if
+        ``coerce_year`` is not specified, but ``roll_utc_offset`` is specified.
```
Understandable. I have a slight preference for the default behavior. I believe folks familiar with PVsyst will assume 1990. Maybe we should leave this PR open for a few days to invite more feedback?

mikofski replied:
Since there's no further feedback, should we merge this PR?

mikofski commented:
> One thing I noticed was that the index name is maintained as "time(UTC)" when utc_roll_offset is None and is removed when utc_roll_offset is specified. Should this be tested? Maybe it doesn't matter

I renamed the index and added tests in https://github.com/pvlib/pvlib-python/pull/2138/commits/1e9ee64186eb74457f1b2769093b406dc39af7a7

AdamRJensen commented:
Thanks @mikofski Great to see this new feature in the iotools!

NikosAlexandris commented:
> > My perception is that this is fixing a flaw in the PVGIS service. Hopefully other TMY sources align on local time.
> 
> The PVGIS team is currently reworking the entire process, so this is a good time to make requests about features and provide feedback. Tagging @NikosAlexandris

Thank you for pinging me here. Will look into this as time permits.

NikosAlexandris commented:
> Tested in this Colab notebook: [`test_coerce_pvgis.ipynb`](https://colab.research.google.com/drive/1pibVTd02QUEJ45oShWuaqzywqviaxWFo?usp=sharing).

@mikofski Would you have some time to help me better understand the "issue" described in this Notebook ?  Thank you.

mikofski commented:
Hi @NikosAlexandris , in past versions of ‘iotools.get_pvgis_tmy()` the data-frame returned was always in UTC time. In my experience TMY files have traditionally been in local time, starting & ending at midnight. Therefore, the “issue” tested in the notebook is if the additional arguments roll utc offset and coerce year will rotate the data frame to coincide with local time, starting at midnight in the desired year
